Add the option to use Workload Identity Federation

daxelrod-rh opened the pull request:
Since Workload Identity Federation is the preferred auth type for OSD-GCP clusters, add the ability to use it. Behavior still defaults to Service Account for backward compatibility, but new projects that copy `terraform.tfvars.example` will get WIF. This behavior is controlled by a new `gcp_authentication_type` variable.

Add a new shell resource to manage the wif-config. This includes a "read" script (which saves the JSON-ified output of `ocm gcp wif-config describe` in Terraform state) to mostly make Terraform behave the way users would expect (in other words, preventing Terraform from wanting to delete/create the wif-config on every run).

Improve documentation for both the WIF and SA options.

daxelrod-rh committed:
Add Workload Identity Configuration shell resource

Add new gcp_authentication_type variable, which defaults to
"service_account" for backward compatability, but allows for specifying
"workload_identity_federation" to use WIF.

When using WIF, a new shell script resource controls running the ocm
commands to create or delete the wif-config. This includes a "read"
action which does a "describe" and then manually converts to JSON, since
that is what the shell provider needs.

Note that the resource is not actually used to alter the behavior of
cluster creation yet.

daxelrod-rh committed:
Wire up wif-config resource to cluster creation

Actually use the wif-config resource to alter which flags are passed to
cluster creation.

Document the new usage.

daxelrod-rh commented:
Note that I was only able to get this to run with the manual addition of `--version 4.17.12` in the cluster create command, because of an OCM bug where not specifying the version causes OCM to not figure out any valid version. It does make sense to add the ability to specify the cluster version to this codebase, but that will be a separate PR.

andyrepton commented on README.md:
Recommended changes:

1. Change this bit:

* Copy and modify the tfvars file in order to custom to your scenario

```bash
cp -pr terraform.tfvars.example terraform.tfvars
```

To this:

* Copy and modify the tfvars file in order to custom to your scenario (note this is in the `configuration/tfvars` sub-folder)

```bash
cp -pr configuration/tfvars/terraform.tfvars.example configuration/tfvars/terraform.tfvars
```

andyrepton commented on README.md:
`ocm`, if using the homebrew version doesn't have `ocm gcp` yet. It would be good to specify a version of `ocm` you need to have installed for this to work in the README

diana-sari committed:
Make osd_gcp_private work

Actually use the osd_gcp_private variable to pass the --private flag to the cluster creation command if true.

This never previously worked, and the variable only controlled whether to create a bastion and its subnet.

daxelrod-rh committed:
Clarify setup instructions

Fix the instructions for copying the tfvars example to list the complete
path relative to the root of the project.

Note the minimum ocm CLI version needed. This is 1.0.3, which is the
version that introduced "ocm gcp verify wif-config".

daxelrod-rh commented:
@andyrepton I fixed both of your suggestions. (Sorry about the wait, a ton of things came up since I initially posted this review). Mind giving another look?

I've also merged in @diana-sari's bugfix for passing the --private flag during cluster creation. This was broken before this PR, but we might as well fix it.
